macOS build requires PkgConfig, cairo, fontconfig and more?

I attempted to follow the [installation instructions](https://github.com/AUSAXS/AUSAXS/wiki/Compilation-&-installation). After installing the required cmake, I got:

```
% cmake -B build -S . -DCMAKE_BUILD_TYPE=Release
[...]
-- building elements with cocoa host UI library
CMake Error at /usr/local/dev/cmake/3.21.4/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE)
Call Stack (most recent call first):
  /usr/local/dev/cmake/3.21.4/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/dev/cmake/3.21.4/share/cmake-3.21/Modules/FindPkgConfig.cmake:70 (find_package_handle_standard_args)
  build/_deps/elements-src/lib/CMakeLists.txt:204 (find_package)

-- Configuring incomplete, errors occurred!
```
The named generated(?) CMakeLists.txt contains:
```
find_package(PkgConfig REQUIRED)
```

After installing pkg-config, I got:
```
--   No package 'cairo' found
CMake Error at /usr/local/dev/cmake/3.21.4/share/cmake-3.21/Modules/FindPkgConfig.cmake:562 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/local/dev/cmake/3.21.4/share/cmake-3.21/Modules/FindPkgConfig.cmake:784 (_pkg_check_modules_internal)
  build/_deps/elements-src/lib/CMakeLists.txt:242 (pkg_check_modules)
```
And I would suspect that all these would fail:
```
pkg_check_modules(cairo REQUIRED IMPORTED_TARGET cairo)
pkg_check_modules(fontconfig REQUIRED IMPORTED_TARGET fontconfig)
pkg_check_modules(freetype2 REQUIRED IMPORTED_TARGET freetype2)
pkg_check_modules(libwebp REQUIRED IMPORTED_TARGET libwebp)
pkg_check_modules(GTK3 REQUIRED IMPORTED_TARGET gtk+-3.0)
```

Ideally it should notice that any/all of these packages are missing and either fail gracefully or build whatever it can without these. Alternatively, please document that everything is required. Thanks.

I can confirm that `-DGUI=OFF` bypasses the cmake error. However, on `make`, I get:
```
 89%] Building CXX object source/md/CMakeFiles/ausaxs_md.dir/utility/files/MDPCreator.cpp.o
/Users/biosaxs/git/AUSAXS/source/md/utility/files/MDPCreator.cpp:20:23: error: implicit instantiation of undefined template 'std::basic_stringstream<char>'
   20 |     std::stringstream ss;
      |                       ^
/Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk/usr/include/c++/v1/__fwd/sstream.h:29:28: note: template is declared here
   29 | class _LIBCPP_TEMPLATE_VIS basic_stringstream;
      |                            ^
1 error generated.
```
This helps:
```diff
% git diff
diff --git a/source/md/utility/files/MDPCreator.cpp b/source/md/utility/files/MDPCreator.cpp
index 53d93e6f..e44de7a3 100644
--- a/source/md/utility/files/MDPCreator.cpp
+++ b/source/md/utility/files/MDPCreator.cpp
@@ -7,6 +7,7 @@ For more information, please refer to the LICENSE file in the project root.
 #include <md/utility/Exceptions.h>
 
 #include <iomanip>
+#include <sstream>
 
 using namespace md;
 
```

--- a/executable/md/frame_analysis.cpp
+++ b/executable/md/frame_analysis.cpp
@@ -37,7 +37,7 @@ int main(int argc, char const *argv[]) {
     };
 
     if (io::Folder tmp("temp/md"); !tmp.exists()) {tmp.create();}
-    gmx::gmx::set_logfile(sele.output + "output.log", sele.output + "cmd.log");
+    gmx::gmx::set_logfile(sele.output.str() + "output.log", sele.output.str() + "cmd.log");
     PDBFile pdb(s_pdb);
 
     // prepare sims
@@ -59,7 +59,7 @@ int main(int argc, char const *argv[]) {
     for (unsigned int i = saxs.size(); i > 0; i--) { // reverse wait order since last job is the shortest one
         saxs[i].job->wait();
         auto res = saxs[i].job->result();
-        res.xvg.rename("saxs_" + std::to_string(i) + ".xvg").copy({sele.output + "output"});
+        res.xvg.rename("saxs_" + std::to_string(i) + ".xvg").copy({sele.output.str() + "output"});
     }
 
     return 0;

--- a/executable/md/simulate.cpp
+++ b/executable/md/simulate.cpp
@@ -52,7 +52,7 @@ int main(int argc, char const *argv[]) {
     };
 
     if (io::Folder tmp("temp/md"); !tmp.exists()) {tmp.create();}
-    gmx::gmx::set_logfile(sele.output + "output.log", sele.output + "cmd.log");
+    gmx::gmx::set_logfile(sele.output.str() + "output.log", sele.output.str() + "cmd.log");
     PDBFile pdb(s_pdb);
 
     // prepare sims

--- a/executable/md/simulate_buffer.cpp
+++ b/executable/md/simulate_buffer.cpp
@@ -41,14 +41,14 @@ int main(int argc, char const *argv[]) {
     };
 
     if (io::Folder tmp("temp/md"); !tmp.exists()) {tmp.create();}
-    gmx::gmx::set_logfile(sele.output + "output.log", sele.output + "cmd.log");
+    gmx::gmx::set_logfile(sele.output.str() + "output.log", sele.output.str() + "cmd.log");
     PDBFile pdb(s_pdb);
 
-    GROFile conf(sele.output + "setup/conf.gro");
+    GROFile conf(sele.output.str() + "setup/conf.gro");
         if (!conf.exists()) {
         // prepare the pdb file for gromacs
         auto[conf, _, posre] = pdb2gmx(pdb)
-            .output({sele.output + "setup/"})
+            .output({sele.output.str() + "setup/"})
             .ignore_hydrogens()
             .water_model(sele.watermodel)
             .forcefield(sele.forcefield)
@@ -57,7 +57,7 @@ int main(int argc, char const *argv[]) {
 
     // create a box around the protein
     auto[uc] = editconf(conf)
-        .output(sele.output + "setup/uc.gro")
+        .output(sele.output.str() + "setup/uc.gro")
         .box_type(sele.boxtype)
         .extend(1)
     .run();

--- a/executable/md/time_analysis.cpp
+++ b/executable/md/time_analysis.cpp
@@ -37,7 +37,7 @@ int main(int argc, char const *argv[]) {
     };
 
     if (io::Folder tmp("temp/md"); !tmp.exists()) {tmp.create();}
-    gmx::gmx::set_logfile(sele.output + "output.log", sele.output + "cmd.log");
+    gmx::gmx::set_logfile(sele.output.str() + "output.log", sele.output.str() + "cmd.log");
     PDBFile pdb(s_pdb);
 
     // prepare sims
@@ -55,11 +55,11 @@ int main(int argc, char const *argv[]) {
     SAXSOptions so(saxs_sele, std::move(molecule), std::move(buffer), pdb);
     auto saxs = timeanalysis(so, 500);
 
-    for (unsigned int i = 0; i < saxs.size(); i++) {saxs[i].job->submit();}
-    for (unsigned int i = saxs.size()-1; i >= 0; i--) { // reverse wait order since the last job is the shortest run
+    for (int i = 0; i < static_cast<int>(saxs.size()); i++) {saxs[i].job->submit();}
+    for (int i = static_cast<int>(saxs.size())-1; i >= 0; i--) { // reverse wait order since the last job is the shortest run
         saxs[i].job->wait();
         auto res = saxs[i].job->result();
-        res.xvg.rename(std::to_string(i+1) + ".xvg").copy({sele.output + "output"});
+        res.xvg.rename(std::to_string(i+1) + ".xvg").copy({sele.output.str() + "output"});
     }
     return 0;
 } 

--- a/include/md/md/programs/mdrun/Execution.h
+++ b/include/md/md/programs/mdrun/Execution.h
@@ -72,8 +72,8 @@ namespace ausaxs::md {
             ~SmaugExecution() override = default;
 
             SmaugExecution(const TPRFile& tpr, const std::string& folder, const std::string& name, const SHFile& jobscript) : folder(folder) {
-                this->filename = tpr.directory() + "/job.sh";
-                this->cmd.set("cd " + tpr.directory() + "; " + jobscript.path() + " -j " + name + " -version release-2021.swaxs -gpu-gener any+ampere -tpr " + tpr.absolute_path());
+                this->filename = tpr.directory().str() + "/job.sh";
+                this->cmd.set("cd " + tpr.directory().str() + "; " + jobscript.path() + " -j " + name + " -version release-2021.swaxs -gpu-gener any+ampere -tpr " + tpr.absolute_path());
             }
 
             SmaugExecution(const std::string& args, const std::string& _export, const std::string& folder, const std::string& name, const SHFile& jobscript) : folder(folder) {

--- a/include/md/md/programs/pdb2gmx.h
+++ b/include/md/md/programs/pdb2gmx.h
@@ -23,9 +23,9 @@ namespace ausaxs::md {
 
             pdb2gmx& output(const io::Folder& path) {
                 folder = path;
-                options.push_back(std::make_shared<shell::Argument>("-o", path + "conf.gro"));
-                options.push_back(std::make_shared<shell::Argument>("-p", path + "topol.top"));
-                options.push_back(std::make_shared<shell::Argument>("-i", path + "posre.itp"));
+                options.push_back(std::make_shared<shell::Argument>("-o", path.str() + "conf.gro"));
+                options.push_back(std::make_shared<shell::Argument>("-p", path.str() + "topol.top"));
+                options.push_back(std::make_shared<shell::Argument>("-i", path.str() + "posre.itp"));
                 return *this;
             }
 
@@ -57,10 +57,10 @@ namespace ausaxs::md {
 
             std::tuple<GROFile, TOPFile, ITPFile> run() {
                 execute();
-                TOPFile top(folder + "topol.top");
+                TOPFile top(folder.str() + "topol.top");
                 top.extract_single_chain();
                 top.fix_relative_includes();
-                return std::make_tuple(GROFile(folder + "conf.gro"), top, ITPFile(folder + "posre.itp"));
+                return std::make_tuple(GROFile(folder.str() + "conf.gro"), top, ITPFile(folder.str() + "posre.itp"));
             }
 
         private: 

--- a/source/crystal/CrystalScattering.cpp
+++ b/source/crystal/CrystalScattering.cpp
@@ -48,8 +48,8 @@ void CrystalScattering::random_rotation() {
     std::mt19937 gen(rd());
     std::uniform_real_distribution<> dis(0, 1);
     double theta = std::acos(2*dis(gen) - 1);
-    double phi = 2*constants::pi*dis(gen);
-    double psi = 2*constants::pi*dis(gen);
+    double phi = 2*std::numbers::pi*dis(gen);
+    double psi = 2*std::numbers::pi*dis(gen);
     auto v = Vector3<double>(std::cos(theta)*std::cos(phi), std::sin(theta)*std::cos(phi), std::sin(phi));
     rotate(v, psi);
 }

--- a/source/crystal/Fval.cpp
+++ b/source/crystal/Fval.cpp
@@ -6,7 +6,8 @@ For more information, please refer to the LICENSE file in the project root.
 #include <crystal/Fval.h>
 #include <utility/Basis3D.h>
 #include <data/Molecule.h>
-#include <data/record/Atom.h>
+#include <data/Body.h>
+#include <data/atoms/AtomFF.h>
 #include <constants/Constants.h>
 
 using namespace ausaxs;
@@ -24,9 +25,9 @@ void Fval::set_points(std::vector<Vector3<double>>&& points) {
 }
 
 void Fval::set_basis(const Basis3D& basis) {
-    ap = {2*constants::pi/basis.x.x(), 0, 0};
-    bp = {0, 2*constants::pi/basis.y.y(), 0};
-    cp = {0, 0, 2*constants::pi/basis.z.z()};
+    ap = {2*std::numbers::pi/basis.x.x(), 0, 0};
+    bp = {0, 2*std::numbers::pi/basis.y.y(), 0};
+    cp = {0, 0, 2*std::numbers::pi/basis.z.z()};
 }
 
 std::vector<Vector3<double>>& Fval::get_points() {
@@ -42,11 +43,11 @@ double Fval::I() const {
 }
 
 data::Molecule Fval::as_protein() {
-    std::vector<data::record::Atom> atoms(points.size());
+    std::vector<data::Atom> atoms(points.size());
     for (unsigned int i = 0; i < points.size(); i++) {
-        atoms[i] = data::record::Atom(points[i], 1, constants::atom_t::H, "LYS", i);
+        atoms[i] = data::Atom(points[i], 1);
     } 
-    return data::Molecule(atoms);
+    return data::Molecule(std::vector{data::Body(atoms)});
 }
 
 // #include <settings/CrystalSettings.h>
@@ -60,9 +61,9 @@ data::Molecule Fval::as_protein() {
 //     z_factors.reserve(points.size());
 
 //     for (const Vector3<double>& point : points) {
-//         std::complex<double> x_factor = std::polar(1.0, -2*constants::pi*ap.x()*point.x());
-//         std::complex<double> y_factor = std::polar(1.0, -2*constants::pi*bp.y()*point.y());
-//         std::complex<double> z_factor = std::polar(1.0, -2*constants::pi*cp.z()*point.z());
+//         std::complex<double> x_factor = std::polar(1.0, -2*std::numbers::pi*ap.x()*point.x());
+//         std::complex<double> y_factor = std::polar(1.0, -2*std::numbers::pi*bp.y()*point.y());
+//         std::complex<double> z_factor = std::polar(1.0, -2*std::numbers::pi*cp.z()*point.z());
 
 //         x_factors.push_back(x_factor);
 //         y_factors.push_back(y_factor);

--- a/source/crystal/io/CrystalReaderFactory.cpp
+++ b/source/crystal/io/CrystalReaderFactory.cpp
@@ -23,7 +23,7 @@ namespace ausaxs::crystal::factory {
         } else if (constants::filetypes::structure.check(filename)) {
             return std::make_unique<crystal::io::PDBReader>();
         } else {
-            throw except::io_error("crystal::io::CrystalReaderFactory::create: Unknown file extension for file \"" + filename + "\"");
+            throw except::io_error("crystal::io::CrystalReaderFactory::create: Unknown file extension for file \"" + filename.str() + "\"");
         }
     }
 }

--- a/source/crystal/io/PDBReader.cpp
+++ b/source/crystal/io/PDBReader.cpp
@@ -5,14 +5,12 @@ For more information, please refer to the LICENSE file in the project root.
 
 #include <crystal/io/PDBReader.h>
 #include <data/Molecule.h>
+#include <data/Body.h>
 #include <utility/Axis3D.h>
 #include <utility/Basis3D.h>
 #include <settings/CrystalSettings.h>
 #include <settings/HistogramSettings.h>
 #include <io/ExistingFile.h>
-#include <data/record/Atom.h>
-#include <data/record/Water.h>
-#include <data/Body.h>
 #include <constants/Constants.h>
 
 using namespace ausaxs;
@@ -23,10 +21,10 @@ std::pair<Basis3D, std::vector<Vector3<double>>> crystal::io::PDBReader::read(co
 
     auto prot_atoms = protein.get_atoms();
     if (prot_atoms.empty()) {throw except::invalid_argument("PDBReader::read: No atoms were found in file \"" + input.str() + "\".");}
-    auto position = prot_atoms[0].get_coordinates();
+    auto position = prot_atoms[0].coordinates();
     Axis3D axis({position.x(), position.x(), position.y(), position.y(), position.z(), position.z()});
     for (const auto& atom : prot_atoms) {
-        auto position = atom.get_coordinates();
+        auto position = atom.coordinates();
         if (position.x() < axis.x.min) {axis.x.min = position.x();}
         if (position.x() > axis.x.max) {axis.x.max = position.x();}
         if (position.y() < axis.y.min) {axis.y.min = position.y();}
@@ -46,7 +44,7 @@ std::pair<Basis3D, std::vector<Vector3<double>>> crystal::io::PDBReader::read(co
     protein.translate({-axis.x.min + factor*axis.x.span(), -axis.y.min + factor*axis.y.span(), -axis.z.min + factor*axis.z.span()});
     std::vector<Vector3<double>> atoms;
     for (const auto& atom : prot_atoms) {
-        atoms.push_back(atom.get_coordinates());
+        atoms.push_back(atom.coordinates());
     }
 
     // expand the box by 2 times

--- a/source/crystal/miller/FibonacciMillers.cpp
+++ b/source/crystal/miller/FibonacciMillers.cpp
@@ -7,7 +7,6 @@ For more information, please refer to the LICENSE file in the project root.
 #include <crystal/miller/Miller.h>
 #include <crystal/Fval.h>
 #include <settings/CrystalSettings.h>
-#include <constants/ConstantsMath.h>
 
 #include <math.h>
 #include <iostream>
@@ -129,7 +128,7 @@ std::vector<ausaxs::Vector3<double>> FibonacciMillers::generate_fibonacci_sphere
     std::vector<Vector3<double>> points(n);
 
     for (int i = 0; i < n; i++) {
-        double theta = 2*constants::pi*i/phi;
+        double theta = 2*std::numbers::pi*i/phi;
         double z = 1 - 2*(i + 0.5)/n;
         double r = sqrt(1 - z*z);
         points[i] = Vector3<double>(r*cos(theta), r*sin(theta), z);

--- a/source/crystal/miller/MillerGenerationFactory.cpp
+++ b/source/crystal/miller/MillerGenerationFactory.cpp
@@ -8,7 +8,6 @@ For more information, please refer to the LICENSE file in the project root.
 #include <crystal/miller/ReducedMillers.h>
 #include <crystal/miller/FibonacciMillers.h>
 #include <settings/CrystalSettings.h>
-#include <residue/ResidueMap.h>
 
 using namespace ausaxs;
 

--- a/source/md/programs/gmx.cpp
+++ b/source/md/programs/gmx.cpp
@@ -43,7 +43,7 @@ std::string gmx::execute() {
         // this will crash if using a shell script, so run everything in bash and encapsulate entire cmd with ''
         cmd.set(escape_single_quotes(cmd.get()));
         cmd.prepend("exec bash -c 'set -o pipefail; ");
-        cmd.append("2>&1 | tee -a " + outputlog + "'");
+        cmd.append("2>&1 | tee -a " + outputlog.str() + "'");
     }
 
     auto result = cmd.execute();

--- a/source/md/programs/mdrun.cpp
+++ b/source/md/programs/mdrun.cpp
@@ -26,7 +26,7 @@ mdrun& mdrun::input(const TPRFile& tpr) {
 
 mdrun& mdrun::output(const io::Folder& folder, const std::string& prefix) {
     this->folder = folder;
-    options.push_back(std::make_shared<shell::Argument>("-deffnm", folder + prefix));
+    options.push_back(std::make_shared<shell::Argument>("-deffnm", folder.str() + prefix));
     return *this;
 }
 

--- a/source/md/simulate/buffer.cpp
+++ b/source/md/simulate/buffer.cpp
@@ -16,11 +16,11 @@ md::SimulateBufferOutput md::simulate_buffer(const BufferOptions& options) {
     //##################################//
     //###           GLOBALS          ###//
     //##################################//
-    io::Folder mdp_folder = options.output + "mdp/";
-    io::Folder setup_path = options.output + "buffer/setup/";
-    io::Folder em_path = options.output + "buffer/em/";
-    io::Folder eq_path = options.output + "buffer/eq/";
-    io::Folder prod_path = options.output + "buffer/prod/";
+    io::Folder mdp_folder = options.output.str() + "mdp/";
+    io::Folder setup_path = options.output.str() + "buffer/setup/";
+    io::Folder em_path = options.output.str() + "buffer/em/";
+    io::Folder eq_path = options.output.str() + "buffer/eq/";
+    io::Folder prod_path = options.output.str() + "buffer/prod/";
     mdp_folder.create(); setup_path.create(); em_path.create(); eq_path.create(); prod_path.create();
 
     if (!options.refgro.exists()) {
@@ -30,8 +30,8 @@ md::SimulateBufferOutput md::simulate_buffer(const BufferOptions& options) {
     //##################################//
     //###           SETUP            ###//
     //##################################//
-    GROFile gro(setup_path + "buffer.gro");
-    TOPFile top(setup_path + "topol.top");
+    GROFile gro(setup_path.str() + "buffer.gro");
+    TOPFile top(setup_path.str() + "topol.top");
     if (!gro.exists() || !top.exists()) {
         {
             std::cout << "\tCopying unit cell from molecule..." << std::flush;
@@ -64,7 +64,7 @@ md::SimulateBufferOutput md::simulate_buffer(const BufferOptions& options) {
         }
 
         auto[ucgro] = editconf(options.refgro)
-            .output(setup_path + "uc.gro")
+            .output(setup_path.str() + "uc.gro")
             .box_type(options.boxtype)
             .extend(2)
         .run();
@@ -94,14 +94,14 @@ md::SimulateBufferOutput md::simulate_buffer(const BufferOptions& options) {
     //##################################//
     //###     ENERGY MINIMIZATION    ###//
     //##################################//
-    GROFile emgro(em_path + "em.gro");
+    GROFile emgro(em_path.str() + "em.gro");
     if (!emgro.exists()) {
         std::cout << "\tRunning energy minimization..." << std::flush;
 
         // prepare energy minimization sim
-        MDPFile mdp = EMMDPCreator().write(mdp_folder + "emsol.mdp");
+        MDPFile mdp = EMMDPCreator().write(mdp_folder.str() + "emsol.mdp");
         auto[emtpr] = grompp(mdp, top, gro)
-            .output(em_path + "em.tpr")
+            .output(em_path.str() + "em.tpr")
             .restraints(gro)
             .warnings(1)
         .run();
@@ -119,14 +119,14 @@ md::SimulateBufferOutput md::simulate_buffer(const BufferOptions& options) {
     //##################################//
     //###       EQUILIBRATION        ###//
     //##################################//
-    GROFile eqgro(eq_path + "eq.gro");
+    GROFile eqgro(eq_path.str() + "eq.gro");
     if (!eqgro.exists()) {
         std::cout << "\tRunning thermalization..." << std::flush;
 
         // prepare equilibration sim
-        MDPFile mdp = EQMDPCreatorSol().write(mdp_folder + "eqsol.mdp");
+        MDPFile mdp = EQMDPCreatorSol().write(mdp_folder.str() + "eqsol.mdp");
         auto[eqtpr] = grompp(mdp, top, emgro)
-            .output(eq_path + "eq.tpr")
+            .output(eq_path.str() + "eq.tpr")
             .restraints(emgro)
             .warnings(1)
         .run();
@@ -144,14 +144,14 @@ md::SimulateBufferOutput md::simulate_buffer(const BufferOptions& options) {
     //##################################//
     //###       PRODUCTION          ###//
     //##################################//
-    GROFile prodgro(prod_path + "confout.gro");
+    GROFile prodgro(prod_path.str() + "confout.gro");
     if (!prodgro.exists()) {
         std::cout << "\tRunning production..." << std::flush;
 
         // prepare production sim
-        MDPFile mdp = options.bufmdp->write(mdp_folder + "prbuf.mdp");
+        MDPFile mdp = options.bufmdp->write(mdp_folder.str() + "prbuf.mdp");
         auto[prodtpr] = grompp(mdp, top, eqgro)
-            .output(prod_path + "prod.tpr")
+            .output(prod_path.str() + "prod.tpr")
             .warnings(1)
         .run();
 

--- a/source/md/simulate/frameanalysis.cpp
+++ b/source/md/simulate/frameanalysis.cpp
@@ -31,14 +31,14 @@ std::vector<md::SAXSOutput> md::frameanalysis(SAXSOptions& options) {
     //##################################//
     //###           GLOBALS          ###//
     //##################################//
-    io::Folder output = options.output + "saxs/";
-    io::Folder protein_path = output + "protein/";
-    io::Folder buffer_path = output + "buffer/";
-    io::Folder mdp_folder = output + "mdp/";
-    io::Folder prod_folder = output + "prod/";
+    io::Folder output = options.output.str() + "saxs/";
+    io::Folder protein_path = output.str() + "protein/";
+    io::Folder buffer_path = output.str() + "buffer/";
+    io::Folder mdp_folder = output.str() + "mdp/";
+    io::Folder prod_folder = output.str() + "prod/";
 
-    TOPFile moltop(protein_path + "topol.top");
-    TOPFile buftop(buffer_path + "topol.top");
+    TOPFile moltop(protein_path.str() + "topol.top");
+    TOPFile buftop(buffer_path.str() + "topol.top");
     if (!moltop.exists() || !buftop.exists()) {
         options.molecule.top.copy(moltop.directory());
         options.buffer.top.copy(buftop.directory());
@@ -49,8 +49,8 @@ std::vector<md::SAXSOutput> md::frameanalysis(SAXSOptions& options) {
     //##################################//
     // we want to always have the group Water_and_ions, but it is only generated automatically if there are actual ions present in the system. 
     // since this is not guaranteed, we have to generate it manually
-    NDXFile molindex(protein_path + "index.ndx");
-    NDXFile bufindex(buffer_path + "index.ndx");
+    NDXFile molindex(protein_path.str() + "index.ndx");
+    NDXFile bufindex(buffer_path.str() + "index.ndx");
 
     if (!molindex.exists()) {
         make_ndx(molgro)
@@ -85,21 +85,21 @@ std::vector<md::SAXSOutput> md::frameanalysis(SAXSOptions& options) {
     //##################################//
     //###         ENVELOPE           ###//
     //##################################//
-    GROFile envgro(protein_path + "envelope-ref.gro");
-    PYFile envpy(protein_path + "envelope.py");
-    DATFile envdat(protein_path + "envelope.dat");
-    MDPFile molmdp(mdp_folder + "rerun_mol.mdp");
+    GROFile envgro(protein_path.str() + "envelope-ref.gro");
+    PYFile envpy(protein_path.str() + "envelope.py");
+    DATFile envdat(protein_path.str() + "envelope.dat");
+    MDPFile molmdp(mdp_folder.str() + "rerun_mol.mdp");
 
     if (!envgro.exists() || !envpy.exists() || !envdat.exists() || !molmdp.exists()) {
-        MDPFile dummymdp = MDPFile(output + "empty.mdp"); dummymdp.create();
+        MDPFile dummymdp = MDPFile(output.str() + "empty.mdp"); dummymdp.create();
         auto[dummytpr] = grompp(dummymdp, moltop, molgro)
-            .output(output + "saxs.tpr")
+            .output(output.str() + "saxs.tpr")
             .warnings(1)
         .run();
         dummymdp.remove();
 
         auto[itps] = genscatt(dummytpr, molindex)
-            .output(protein_path + "scatt.itp")
+            .output(protein_path.str() + "scatt.itp")
             .group("Protein")
         .run();
 
@@ -107,13 +107,13 @@ std::vector<md::SAXSOutput> md::frameanalysis(SAXSOptions& options) {
 
         // dump the first 50 frames
         auto[traj] = trjconv(molxtc)
-            .output(protein_path + "protein.xtc")
+            .output(protein_path.str() + "protein.xtc")
             .startframe(50)
         .run();
 
         // center the trajectories
         auto[cluster] = trjconv(traj)
-            .output(protein_path + "cluster.xtc")
+            .output(protein_path.str() + "cluster.xtc")
             .group("Protein")
             .pbc("cluster")
             .ur("tric")
@@ -122,7 +122,7 @@ std::vector<md::SAXSOutput> md::frameanalysis(SAXSOptions& options) {
         .run();
 
         auto[centered] = trjconv(cluster)
-            .output(protein_path + "centered.xtc")
+            .output(protein_path.str() + "centered.xtc")
             .group("Protein")
             .center()
             .boxcenter("tric")
@@ -154,20 +154,20 @@ std::vector<md::SAXSOutput> md::frameanalysis(SAXSOptions& options) {
         }
     }
 
-    MDPFile bufmdp(mdp_folder + "rerun_buf.mdp");
+    MDPFile bufmdp(mdp_folder.str() + "rerun_buf.mdp");
     if (!bufmdp.exists()) {
         SAXSMDPCreatorSol().write(bufmdp);
     }
 
     // generate run files - they can be reused for all parts
     auto[moltpr] = grompp(molmdp, moltop, molgro)
-        .output(prod_folder + "mol.tpr")
+        .output(prod_folder.str() + "mol.tpr")
         .index(molindex)
         .warnings(1)
     .run();
 
     auto[buftpr] = grompp(bufmdp, buftop, bufgro)
-        .output(prod_folder + "buf.tpr")
+        .output(prod_folder.str() + "buf.tpr")
         .index(bufindex)
         .warnings(2)
     .run();
@@ -175,15 +175,15 @@ std::vector<md::SAXSOutput> md::frameanalysis(SAXSOptions& options) {
     // chop the trajectory into separate parts
     std::vector<SAXSOutput> jobs;
     for (unsigned int i = 0; i < 10; ++i) {
-        io::Folder part_folder = prod_folder + "part_" + std::to_string(i) + "/";
+        io::Folder part_folder = prod_folder.str() + "part_" + std::to_string(i) + "/";
 
         auto [molxtc_i] = trjconv(molxtc)
-            .output(part_folder + "mol.xtc")
+            .output(part_folder.str() + "mol.xtc")
             .skip_every_n_frame(i+1)
         .run();
 
         auto [bufxtc_i] = trjconv(bufxtc)
-            .output(part_folder + "buf.xtc")
+            .output(part_folder.str() + "buf.xtc")
             .skip_every_n_frame(i+1)
         .run();
 

--- a/source/md/simulate/molecule.cpp
+++ b/source/md/simulate/molecule.cpp
@@ -26,21 +26,21 @@ SimulateMoleculeOutput md::simulate_molecule(MoleculeOptions& options) {
     //##################################//
     //###           GLOBALS          ###//
     //##################################//
-    io::Folder mdp_folder = options.output + "mdp/";
-    io::Folder setup_path = options.output + "protein/setup/";
-    io::Folder em_path = options.output + "protein/em/";
-    io::Folder eq_path = options.output + "protein/eq/";
-    io::Folder prod_path = options.output + "protein/prod/";
+    io::Folder mdp_folder = options.output.str() + "mdp/";
+    io::Folder setup_path = options.output.str() + "protein/setup/";
+    io::Folder em_path = options.output.str() + "protein/em/";
+    io::Folder eq_path = options.output.str() + "protein/eq/";
+    io::Folder prod_path = options.output.str() + "protein/prod/";
     mdp_folder.create(); setup_path.create(); em_path.create(); eq_path.create(); prod_path.create();
 
     // gmx::gmx::set_outputlog(output + "gmx.log");
     //##################################//
     //###           SETUP            ###//
     //##################################//
-    GROFile solv_ion(setup_path + "solv_ion.gro");
-    TOPFile top(setup_path + "topol.top");
+    GROFile solv_ion(setup_path.str() + "solv_ion.gro");
+    TOPFile top(setup_path.str() + "topol.top");
     if (!solv_ion.exists() || !top.exists()) {
-        GROFile conf(setup_path + "conf.gro");
+        GROFile conf(setup_path.str() + "conf.gro");
         if (!conf.exists()) {
             console::print_text("Converting structure to GROMACS format...");
 
@@ -59,7 +59,7 @@ SimulateMoleculeOutput md::simulate_molecule(MoleculeOptions& options) {
         // create a box around the protein
         console::print_text("Generating unit cell...");
         auto[uc] = editconf(conf)
-            .output(setup_path + "uc.gro")
+            .output(setup_path.str() + "uc.gro")
             .box_type(options.boxtype)
             .extend(1.5)
         .run();
@@ -67,16 +67,16 @@ SimulateMoleculeOutput md::simulate_molecule(MoleculeOptions& options) {
         // add water to the box
         console::print_text("Solvating unit cell...");
         auto[solv] = solvate(uc)
-            .output(setup_path + "solv.gro")
+            .output(setup_path.str() + "solv.gro")
             .solvent(options.forcefield, options.watermodel)
             .radius(0.2)
             .topology(top)
         .run();
 
         // generate an empty tpr file 
-        MDPFile mdp = MDPFile(setup_path + "empty.mdp"); mdp.create();
+        MDPFile mdp = MDPFile(setup_path.str() + "empty.mdp"); mdp.create();
         auto[ions] = grompp(mdp, top, solv)
-            .output(setup_path + "ions.tpr")
+            .output(setup_path.str() + "ions.tpr")
             .warnings(1)
         .run();
         mdp.remove();
@@ -98,14 +98,14 @@ SimulateMoleculeOutput md::simulate_molecule(MoleculeOptions& options) {
     //##################################//
     //###     ENERGY MINIMIZATION    ###//
     //##################################//
-    GROFile emgro(em_path + "em.gro");
+    GROFile emgro(em_path.str() + "em.gro");
     if (!emgro.exists()) {
         console::print_text("Running energy minimization...");
 
         // prepare energy minimization sim
-        MDPFile mdp = EMMDPCreator().write(mdp_folder + "emmol.mdp");
+        MDPFile mdp = EMMDPCreator().write(mdp_folder.str() + "emmol.mdp");
         auto[emtpr] = grompp(mdp, top, solv_ion)
-            .output(em_path + "em.tpr")
+            .output(em_path.str() + "em.tpr")
             .restraints(solv_ion)
         .run();
 
@@ -121,8 +121,8 @@ SimulateMoleculeOutput md::simulate_molecule(MoleculeOptions& options) {
     //##################################//
     //###       EQUILIBRATION        ###//
     //##################################//
-    GROFile eqgro(eq_path + "eq.gro");
-    NDXFile index(eq_path + "index.ndx");
+    GROFile eqgro(eq_path.str() + "eq.gro");
+    NDXFile index(eq_path.str() + "index.ndx");
     if (!eqgro.exists()) {
         console::print_text("Running thermalization...");
 
@@ -163,9 +163,9 @@ SimulateMoleculeOutput md::simulate_molecule(MoleculeOptions& options) {
         // }
 
         // prepare equilibration sim
-        MDPFile mdp = EQMDPCreatorMol().write(mdp_folder + "eqmol.mdp");
+        MDPFile mdp = EQMDPCreatorMol().write(mdp_folder.str() + "eqmol.mdp");
         auto[eqtpr] = grompp(mdp, top, emgro)
-            .output(eq_path + "eq.tpr")
+            .output(eq_path.str() + "eq.tpr")
             .index(index)
             .restraints(emgro)
             .warnings(1)
@@ -183,11 +183,11 @@ SimulateMoleculeOutput md::simulate_molecule(MoleculeOptions& options) {
     //##################################//
     //###       PRODUCTION           ###//
     //##################################//
-    GROFile prodgro(prod_path + "prod.gro");
+    GROFile prodgro(prod_path.str() + "prod.gro");
     if (!prodgro.exists()) {
         console::print_text("Generating backbone restraints...");
-        auto[backbone] = genrestr(eq_path + "eq.gro")
-            .output(setup_path + "backbone.itp")
+        auto[backbone] = genrestr(eq_path.str() + "eq.gro")
+            .output(setup_path.str() + "backbone.itp")
             .index(index)
             .force(2000, 2000, 2000)
         .run();
@@ -197,9 +197,9 @@ SimulateMoleculeOutput md::simulate_molecule(MoleculeOptions& options) {
 
         // prepare production sim
         console::print_text("Running production...");
-        MDPFile mdp = options.molmdp->write(mdp_folder + "prmol.mdp");
+        MDPFile mdp = options.molmdp->write(mdp_folder.str() + "prmol.mdp");
         auto[prodtpr] = grompp(mdp, top, eqgro)
-            .output(prod_path + "prod.tpr")
+            .output(prod_path.str() + "prod.tpr")
             .index(index)
             .restraints(eqgro)
             .warnings(1)

--- a/source/md/simulate/saxs.cpp
+++ b/source/md/simulate/saxs.cpp
@@ -36,15 +36,15 @@ md::SAXSOutput md::simulate_saxs(md::SAXSOptions& options) {
     //##################################//
     //###           GLOBALS          ###//
     //##################################//
-    io::Folder output = options.output + "saxs/";
-    io::Folder protein_path = output + "protein/";
-    io::Folder buffer_path = output + "buffer/";
-    io::Folder mdp_folder = output + "mdp/";
-    io::Folder prod_folder = output + "prod/";
+    io::Folder output = options.output.str() + "saxs/";
+    io::Folder protein_path = output.str() + "protein/";
+    io::Folder buffer_path = output.str() + "buffer/";
+    io::Folder mdp_folder = output.str() + "mdp/";
+    io::Folder prod_folder = output.str() + "prod/";
     mdp_folder.create(); prod_folder.create();
 
-    TOPFile moltop(protein_path + "topol.top");
-    TOPFile buftop(buffer_path + "topol.top");
+    TOPFile moltop(protein_path.str() + "topol.top");
+    TOPFile buftop(buffer_path.str() + "topol.top");
     if (!moltop.exists() || !buftop.exists()) {
         options.molecule.top.copy(moltop.directory());
         options.buffer.top.copy(buftop.directory());
@@ -66,7 +66,7 @@ md::SAXSOutput md::simulate_saxs(md::SAXSOptions& options) {
     //##################################//
     // we want to always have the group Water_and_ions, but it is only generated automatically if there are actual ions present in the system. 
     // since this is not guaranteed, we have to generate it manually
-    NDXFile molindex(protein_path + "index.ndx");
+    NDXFile molindex(protein_path.str() + "index.ndx");
     if (!molindex.exists()) {
         console::print_text("Creating index file for molecule...");
         make_ndx(molgro)
@@ -108,7 +108,7 @@ md::SAXSOutput md::simulate_saxs(md::SAXSOptions& options) {
         console::print_text("Reusing index file for molecule.");
     }
 
-    NDXFile bufindex(buffer_path + "index.ndx");
+    NDXFile bufindex(buffer_path.str() + "index.ndx");
     if (!bufindex.exists()) {
         console::print_text("Creating index file for buffer...");
         make_ndx(bufgro)
@@ -130,24 +130,24 @@ md::SAXSOutput md::simulate_saxs(md::SAXSOptions& options) {
     //##################################//
     //###         ENVELOPE           ###//
     //##################################//
-    GROFile envgro(protein_path + "envelope-ref.gro");
-    PYFile envpy(protein_path + "envelope.py");
-    DATFile envdat(protein_path + "envelope.dat");
-    MDPFile molmdp(mdp_folder + "rerun_mol.mdp");
+    GROFile envgro(protein_path.str() + "envelope-ref.gro");
+    PYFile envpy(protein_path.str() + "envelope.py");
+    DATFile envdat(protein_path.str() + "envelope.dat");
+    MDPFile molmdp(mdp_folder.str() + "rerun_mol.mdp");
 
     if (!envgro.exists() || !envpy.exists() || !envdat.exists() || !molmdp.exists()) {
         console::print_text("Preparing SAXS rerun...");
         console::indent();
-        MDPFile dummymdp = MDPFile(output + "empty.mdp"); dummymdp.create();
+        MDPFile dummymdp = MDPFile(output.str() + "empty.mdp"); dummymdp.create();
         auto[dummytpr] = grompp(dummymdp, moltop, molgro)
-            .output(output + "saxs.tpr")
+            .output(output.str() + "saxs.tpr")
             .warnings(1)
         .run();
         dummymdp.remove();
 
         console::print_text("Generating scattering parameters...");
         auto[itps] = genscatt(dummytpr, molindex)
-            .output(protein_path + "scatt.itp")
+            .output(protein_path.str() + "scatt.itp")
             .group("Protein_and_Ions")
         .run();
 
@@ -165,14 +165,14 @@ md::SAXSOutput md::simulate_saxs(md::SAXSOptions& options) {
         // dump the first 50 frames
         console::print_text("Dumping first 50 frames...");
         auto[traj] = trjconv(molxtc)
-            .output(protein_path + "protein.xtc")
+            .output(protein_path.str() + "protein.xtc")
             .startframe(50)
         .run();
 
         // center the trajectories
         console::print_text("Centering trajectories...");
         auto[cluster] = trjconv(traj)
-            .output(protein_path + "cluster.xtc")
+            .output(protein_path.str() + "cluster.xtc")
             .group("Protein")
             .pbc("cluster")
             .ur("tric")
@@ -181,7 +181,7 @@ md::SAXSOutput md::simulate_saxs(md::SAXSOptions& options) {
         .run();
 
         auto[centered] = trjconv(cluster)
-            .output(protein_path + "centered.xtc")
+            .output(protein_path.str() + "centered.xtc")
             .group("Protein")
             .center()
             .boxcenter("tric")
@@ -218,21 +218,21 @@ md::SAXSOutput md::simulate_saxs(md::SAXSOptions& options) {
         console::print_text("Reusing previously generated envelope.");
     }
 
-    MDPFile bufmdp(mdp_folder + "rerun_buf.mdp");
+    MDPFile bufmdp(mdp_folder.str() + "rerun_buf.mdp");
     if (!bufmdp.exists()) {
         SAXSMDPCreatorSol().write(bufmdp);
     }
 
-    GROFile gro(prod_folder + "prod.gro");
+    GROFile gro(prod_folder.str() + "prod.gro");
     if (!gro.exists()) {
         auto[moltpr] = grompp(molmdp, moltop, molgro)
-            .output(prod_folder + "rerun_mol.tpr")
+            .output(prod_folder.str() + "rerun_mol.tpr")
             .index(molindex)
             .warnings(1)
         .run();
 
         auto[buftpr] = grompp(bufmdp, buftop, bufgro)
-            .output(prod_folder + "rerun_buf.tpr")
+            .output(prod_folder.str() + "rerun_buf.tpr")
             .index(bufindex)
             .warnings(2)
         .run();

--- a/source/md/simulate/timeanalysis.cpp
+++ b/source/md/simulate/timeanalysis.cpp
@@ -31,14 +31,14 @@ std::vector<md::SAXSOutput> md::timeanalysis(SAXSOptions& options, double timest
     //##################################//
     //###           GLOBALS          ###//
     //##################################//
-    io::Folder output = options.output + "saxs/";
-    io::Folder protein_path = output + "protein/";
-    io::Folder buffer_path = output + "buffer/";
-    io::Folder mdp_folder = output + "mdp/";
-    io::Folder prod_folder = output + "prod/";
+    io::Folder output = options.output.str() + "saxs/";
+    io::Folder protein_path = output.str() + "protein/";
+    io::Folder buffer_path = output.str() + "buffer/";
+    io::Folder mdp_folder = output.str() + "mdp/";
+    io::Folder prod_folder = output.str() + "prod/";
 
-    TOPFile moltop(protein_path + "topol.top");
-    TOPFile buftop(buffer_path + "topol.top");
+    TOPFile moltop(protein_path.str() + "topol.top");
+    TOPFile buftop(buffer_path.str() + "topol.top");
     if (!moltop.exists() || !buftop.exists()) {
         options.molecule.top.copy(moltop.directory());
         options.buffer.top.copy(buftop.directory());
@@ -49,8 +49,8 @@ std::vector<md::SAXSOutput> md::timeanalysis(SAXSOptions& options, double timest
     //##################################//
     // we want to always have the group Water_and_ions, but it is only generated automatically if there are actual ions present in the system. 
     // since this is not guaranteed, we have to generate it manually
-    NDXFile molindex(protein_path + "index.ndx");
-    NDXFile bufindex(buffer_path + "index.ndx");
+    NDXFile molindex(protein_path.str() + "index.ndx");
+    NDXFile bufindex(buffer_path.str() + "index.ndx");
 
     if (!molindex.exists()) {
         make_ndx(molgro)
@@ -85,21 +85,21 @@ std::vector<md::SAXSOutput> md::timeanalysis(SAXSOptions& options, double timest
     //##################################//
     //###         ENVELOPE           ###//
     //##################################//
-    GROFile envgro(protein_path + "envelope-ref.gro");
-    PYFile envpy(protein_path + "envelope.py");
-    DATFile envdat(protein_path + "envelope.dat");
-    MDPFile molmdp(mdp_folder + "rerun_mol.mdp");
+    GROFile envgro(protein_path.str() + "envelope-ref.gro");
+    PYFile envpy(protein_path.str() + "envelope.py");
+    DATFile envdat(protein_path.str() + "envelope.dat");
+    MDPFile molmdp(mdp_folder.str() + "rerun_mol.mdp");
 
     if (!envgro.exists() || !envpy.exists() || !envdat.exists() || !molmdp.exists()) {
-        MDPFile dummymdp = MDPFile(output + "empty.mdp"); dummymdp.create();
+        MDPFile dummymdp = MDPFile(output.str() + "empty.mdp"); dummymdp.create();
         auto[dummytpr] = grompp(dummymdp, moltop, molgro)
-            .output(output + "saxs.tpr")
+            .output(output.str() + "saxs.tpr")
             .warnings(1)
         .run();
         dummymdp.remove();
 
         auto[itps] = genscatt(dummytpr, molindex)
-            .output(protein_path + "scatt.itp")
+            .output(protein_path.str() + "scatt.itp")
             .group("Protein")
         .run();
 
@@ -107,13 +107,13 @@ std::vector<md::SAXSOutput> md::timeanalysis(SAXSOptions& options, double timest
 
         // dump the first 50 frames
         auto[traj] = trjconv(molxtc)
-            .output(protein_path + "protein.xtc")
+            .output(protein_path.str() + "protein.xtc")
             .startframe(50)
         .run();
 
         // center the trajectories
         auto[cluster] = trjconv(traj)
-            .output(protein_path + "cluster.xtc")
+            .output(protein_path.str() + "cluster.xtc")
             .group("Protein")
             .pbc("cluster")
             .ur("tric")
@@ -122,7 +122,7 @@ std::vector<md::SAXSOutput> md::timeanalysis(SAXSOptions& options, double timest
         .run();
 
         auto[centered] = trjconv(cluster)
-            .output(protein_path + "centered.xtc")
+            .output(protein_path.str() + "centered.xtc")
             .group("Protein")
             .center()
             .boxcenter("tric")
@@ -154,19 +154,19 @@ std::vector<md::SAXSOutput> md::timeanalysis(SAXSOptions& options, double timest
         }
     }
 
-    MDPFile bufmdp(mdp_folder + "rerun_buf.mdp");
+    MDPFile bufmdp(mdp_folder.str() + "rerun_buf.mdp");
     if (!bufmdp.exists()) {
         SAXSMDPCreatorSol().write(bufmdp);
     }
 
     auto[moltpr] = grompp(molmdp, moltop, molgro)
-        .output(prod_folder + "mol.tpr")
+        .output(prod_folder.str() + "mol.tpr")
         .index(molindex)
         .warnings(1)
     .run();
 
     auto[buftpr] = grompp(bufmdp, buftop, bufgro)
-        .output(prod_folder + "buf.tpr")
+        .output(prod_folder.str() + "buf.tpr")
         .index(bufindex)
         .warnings(2)
     .run();
@@ -179,16 +179,16 @@ std::vector<md::SAXSOutput> md::timeanalysis(SAXSOptions& options, double timest
     std::cout << "\tChopping trajectory into " << frames/framestep << " parts." << std::endl;
     unsigned int part = 0;
     for (unsigned int i = 0; i < frames; i+=framestep) {
-        io::Folder part_folder = prod_folder + "part_" + std::to_string(++part) + "/";
+        io::Folder part_folder = prod_folder.str() + "part_" + std::to_string(++part) + "/";
 
         auto [molxtc_i] = trjconv(molxtc)
-            .output(part_folder + "mol.xtc")
+            .output(part_folder.str() + "mol.xtc")
             .startframe(i)
             .endframe(i+framestep)
         .run();
 
         auto [bufxtc_i] = trjconv(bufxtc)
-            .output(part_folder + "buf.xtc")
+            .output(part_folder.str() + "buf.xtc")
             .startframe(i)
             .endframe(i+framestep)
         .run();

--- a/source/md/utility/files/TOPFile.cpp
+++ b/source/md/utility/files/TOPFile.cpp
@@ -250,8 +250,8 @@ void TOPFile::fix_relative_includes() {
 }
 
 void TOPFile::fix_relative_includes(const io::File& path) {
-    if (!path.exists()) {throw except::io_error("TOPFile::fix_relative_includes: \"" + path + "\" does not exist.");}
-    console::print_text_minor("Fixing relative includes in \"" + path + "\".");
+    if (!path.exists()) {throw except::io_error("TOPFile::fix_relative_includes: \"" + path.str() + "\" does not exist.");}
+    console::print_text_minor("Fixing relative includes in \"" + path.str() + "\".");
     console::indent();
     std::ifstream in(path);
 
@@ -261,10 +261,10 @@ void TOPFile::fix_relative_includes(const io::File& path) {
     while (std::getline(in, line)) {
         if (line.find("#include") != std::string::npos) {
             auto index = line.find("\"");
-            if (index == std::string::npos) {throw except::io_error("TOPFile::fix_relative_includes: \"" + path + "\" contains an include that is not in quotes.");}
+            if (index == std::string::npos) {throw except::io_error("TOPFile::fix_relative_includes: \"" + path.str() + "\" contains an include that is not in quotes.");}
             std::string include = line.substr(index + 1);
             index = include.find("\"");
-            if (index == std::string::npos) {throw except::io_error("TOPFile::fix_relative_includes: \"" + path + "\" contains an include quote which is not terminated.");}
+            if (index == std::string::npos) {throw except::io_error("TOPFile::fix_relative_includes: \"" + path.str() + "\" contains an include quote which is not terminated.");}
             include = include.substr(0, index);
 
             // non-existing files are likely relative to the GROMACS directory and shouldn't be touched
@@ -336,11 +336,11 @@ void TOPFile::extract_single_chain() {
 
     std::string itp;
     for (const auto& line : chain_contents) {itp += line + "\n";}
-    std::ofstream out2(directory() + "/" + newfile);
+    std::ofstream out2(directory().str() + "/" + newfile);
     out2 << itp;
 
     // update includes
-    includes.push_back(ITPFile(directory() + "/" + newfile));
+    includes.push_back(ITPFile(directory().str() + "/" + newfile));
 }
 
 std::string TOPFile::copy(const io::Folder& folder) const {

